File: java/CasinoFairness.java
// Public copy of the game-server algorithm for independent audit.
// Keep this file in lockstep with com.near_reality.content.fairness.CasinoFairness.
package com.near_reality.content.fairness;

import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.HexFormat;

public final class CasinoFairness {

    private static final MessageDigest DIGEST;
    private static final int SERVER_SEED_BYTES = 32;
    private static final int CLIENT_SEED_BYTES = 16;
    private static final int MAX_CLIENT_SEED_BYTES = 128;
    private static final int MINES_TILE_COUNT = 25;
    private static final int MIN_MINES = 1;
    private static final int MAX_MINES = 24;
    private static final long UINT_RANGE = 1L << 32;
    private static final long DICE_OUTCOMES = 10_001L;

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();
    private static final HexFormat HEX = HexFormat.of();

    static {
        try {
            DIGEST = MessageDigest.getInstance("SHA-256");
        } catch (final NoSuchAlgorithmException exception) {
            throw new IllegalStateException("SHA-256 is not available.", exception);
        }
    }

    private CasinoFairness() {
    }

    public static String randomServerSeed() {
        return randomHex(SERVER_SEED_BYTES);
    }

    public static String randomClientSeed() {
        return randomHex(CLIENT_SEED_BYTES);
    }

    public static int[] generateMineTiles(
            final String revealedServerSeed,
            final String clientSeed,
            final long nonce,
            final int mineCount
    ) {
        validateServerSeed(revealedServerSeed);
        validateClientSeed(clientSeed);
        if (mineCount < MIN_MINES || mineCount > MAX_MINES) {
            throw new IllegalArgumentException("The mine count must be between 1 and 24.");
        }
        if (nonce < 0L) {
            throw new IllegalArgumentException("The nonce cannot be negative.");
        }

        final int[] tiles = new int[MINES_TILE_COUNT];
        for (int index = 0; index < tiles.length; index++) {
            tiles[index] = index;
        }

        final Sha256ByteStream random = new Sha256ByteStream(
                revealedServerSeed,
                clientSeed,
                nonce
        );
        for (int index = tiles.length - 1; index > 0; index--) {
            final int swapIndex = random.nextInt(index + 1);
            final int tile = tiles[index];
            tiles[index] = tiles[swapIndex];
            tiles[swapIndex] = tile;
        }

        final int[] mineTiles = Arrays.copyOf(tiles, mineCount);
        Arrays.sort(mineTiles);
        return mineTiles;
    }

    public static DiceRollProof verifyDiceRoll(
            final String serverSeed,
            final String clientSeed,
            final long nonce
    ) {
        validateServerSeed(serverSeed);
        validateClientSeed(clientSeed);
        if (nonce < 0L) {
            throw new IllegalArgumentException("The nonce cannot be negative.");
        }

        final byte[] digest = sha256((clientSeed + ':' + serverSeed + ':' + nonce).getBytes(StandardCharsets.UTF_8));
        final long firstFourBytes = Integer.toUnsignedLong(ByteBuffer.wrap(digest, 0, Integer.BYTES).getInt());
        final int rollBasisPoints = (int) ((firstFourBytes * DICE_OUTCOMES) / UINT_RANGE);
        return new DiceRollProof(
                BigDecimal.valueOf(rollBasisPoints, 2),
                rollBasisPoints,
                hashServerSeed(serverSeed),
                clientSeed,
                nonce,
                HEX.formatHex(digest)
        );
    }

    public static String hashServerSeed(final String serverSeed) {
        validateServerSeed(serverSeed);
        return HEX.formatHex(sha256(serverSeed.getBytes(StandardCharsets.UTF_8)));
    }

    public static boolean verifyServerSeed(
            final String serverSeed,
            final String expectedHash
    ) {
        if (!isValidServerSeed(serverSeed) || expectedHash == null) {
            return false;
        }
        final byte[] actual = hashServerSeed(serverSeed).getBytes(StandardCharsets.US_ASCII);
        final byte[] expected = expectedHash.toLowerCase().getBytes(StandardCharsets.US_ASCII);
        return MessageDigest.isEqual(actual, expected);
    }

    public static void validateClientSeed(final String clientSeed) {
        if (!isValidClientSeed(clientSeed)) {
            throw new IllegalArgumentException(
                    "The client seed must contain 1 to 128 UTF-8 bytes and no control characters."
            );
        }
    }

    public static boolean isValidClientSeed(final String clientSeed) {
        return clientSeed != null
                && !clientSeed.isBlank()
                && clientSeed.getBytes(StandardCharsets.UTF_8).length <= MAX_CLIENT_SEED_BYTES
                && clientSeed.codePoints().noneMatch(Character::isISOControl);
    }

    public static void validateServerSeed(final String serverSeed) {
        if (!isValidServerSeed(serverSeed)) {
            throw new IllegalArgumentException("The server seed must be a 64-character hexadecimal value.");
        }
    }

    public static boolean isValidServerSeed(final String serverSeed) {
        if (serverSeed == null || serverSeed.length() != SERVER_SEED_BYTES * 2) {
            return false;
        }
        try {
            HEX.parseHex(serverSeed);
            return true;
        } catch (final IllegalArgumentException ignored) {
            return false;
        }
    }

    private static byte[] sha256(final byte[] input) {
        return DIGEST.digest(input);
    }

    private static String randomHex(final int byteCount) {
        final byte[] bytes = new byte[byteCount];
        SECURE_RANDOM.nextBytes(bytes);
        return HEX.formatHex(bytes);
    }

    public record DiceRollProof(
            BigDecimal roll,
            int rollBasisPoints,
            String serverSeedHash,
            String clientSeed,
            long nonce,
            String resultDigest
    ) {
    }

    private static final class Sha256ByteStream {

        private final String serverSeed;
        private final String clientSeed;
        private final long nonce;
        private byte[] block = new byte[0];
        private int offset;
        private int blockIndex;

        private Sha256ByteStream(final String serverSeed, final String clientSeed, final long nonce) {
            this.serverSeed = serverSeed;
            this.clientSeed = clientSeed;
            this.nonce = nonce;
        }

        private int nextInt(final int bound) {
            final long rejectionLimit = UINT_RANGE - UINT_RANGE % bound;
            long value;
            do {
                value = nextUnsignedInt();
            } while (value >= rejectionLimit);
            return (int) (value % bound);
        }

        private long nextUnsignedInt() {
            if (offset + Integer.BYTES > block.length) {
                final String message = clientSeed + ':' + serverSeed + ':' + nonce + ':' + blockIndex++;
                block = sha256(message.getBytes(StandardCharsets.UTF_8));
                offset = 0;
            }
            final long value = Integer.toUnsignedLong(ByteBuffer.wrap(block, offset, Integer.BYTES).getInt());
            offset += Integer.BYTES;
            return value;
        }
    }

}

File: public/js/casino-fairness.js
/**
 * Browser/Node port of com.near_reality.content.fairness.CasinoFairness.
 *
 * Hash material is always UTF-8. Dice and mines both use:
 *   SHA-256(clientSeed + ":" + serverSeed + ":" + nonce [+ ":" + blockIndex])
 *
 * Dice maps the first 4 digest bytes with multiply/divide, not modulo:
 *   rollBasisPoints = (uint32 * 10001) / 2^32
 *   roll            = rollBasisPoints / 100
 *
 * Mines Fisher-Yates the 25 tiles with a SHA-256 byte stream and
 * rejection sampling, then takes the first mineCount tiles.
 */

const SERVER_SEED_BYTES = 32;
const MAX_CLIENT_SEED_BYTES = 128;
const MINES_TILE_COUNT = 25;
const MIN_MINES = 1;
const MAX_MINES = 24;
const UINT_RANGE = 1n << 32n;
const DICE_OUTCOMES = 10_001n;
const HEX = /^[0-9a-f]+$/i;

const textEncoder = new TextEncoder();

export const CasinoFairness = {
  SERVER_SEED_BYTES,
  MAX_CLIENT_SEED_BYTES,
  MINES_TILE_COUNT,
  MIN_MINES,
  MAX_MINES,
  DICE_OUTCOMES: Number(DICE_OUTCOMES),

  async sha256(input) {
    const bytes = typeof input === "string" ? textEncoder.encode(input) : input;
    const digest = await crypto.subtle.digest("SHA-256", bytes);
    return new Uint8Array(digest);
  },

  toHex(bytes) {
    return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");
  },

  utf8Length(value) {
    return textEncoder.encode(value).length;
  },

  isIsoControl(codePoint) {
    return codePoint <= 0x1f || (codePoint >= 0x7f && codePoint <= 0x9f);
  },

  isValidClientSeed(clientSeed) {
    return clientSeed != null
      && clientSeed.trim() !== ""
      && this.utf8Length(clientSeed) <= MAX_CLIENT_SEED_BYTES
      && ![...clientSeed].some((char) => this.isIsoControl(char.codePointAt(0)));
  },

  validateClientSeed(clientSeed) {
    if (!this.isValidClientSeed(clientSeed)) {
      throw new Error("The client seed must contain 1 to 128 UTF-8 bytes and no control characters.");
    }
  },

  isValidServerSeed(serverSeed) {
    return typeof serverSeed === "string"
      && serverSeed.length === SERVER_SEED_BYTES * 2
      && HEX.test(serverSeed);
  },

  validateServerSeed(serverSeed) {
    if (!this.isValidServerSeed(serverSeed)) {
      throw new Error("The server seed must be a 64-character hexadecimal value.");
    }
  },

  validateNonce(nonce) {
    if (!Number.isInteger(nonce) || nonce < 0) {
      throw new Error("The nonce cannot be negative.");
    }
  },

  async hashServerSeed(serverSeed) {
    this.validateServerSeed(serverSeed);
    return this.toHex(await this.sha256(serverSeed));
  },

  async verifyServerSeed(serverSeed, expectedHash) {
    if (!this.isValidServerSeed(serverSeed) || expectedHash == null) {
      return false;
    }
    const actual = (await this.hashServerSeed(serverSeed)).toLowerCase();
    return actual === String(expectedHash).toLowerCase();
  },

  async verifyDiceRoll(serverSeed, clientSeed, nonce) {
    this.validateServerSeed(serverSeed);
    this.validateClientSeed(clientSeed);
    this.validateNonce(nonce);

    const digest = await this.sha256(`${clientSeed}:${serverSeed}:${nonce}`);
    const firstFourBytes = unsignedInt32(digest, 0);
    const rollBasisPoints = Number((firstFourBytes * DICE_OUTCOMES) / UINT_RANGE);
    return {
      roll: (rollBasisPoints / 100).toFixed(2),
      rollBasisPoints,
      serverSeedHash: await this.hashServerSeed(serverSeed),
      clientSeed,
      nonce,
      resultDigest: this.toHex(digest)
    };
  },

  async generateMineTiles(revealedServerSeed, clientSeed, nonce, mineCount) {
    this.validateServerSeed(revealedServerSeed);
    this.validateClientSeed(clientSeed);
    this.validateNonce(nonce);
    if (!Number.isInteger(mineCount) || mineCount < MIN_MINES || mineCount > MAX_MINES) {
      throw new Error("The mine count must be between 1 and 24.");
    }

    const tiles = Array.from({ length: MINES_TILE_COUNT }, (_, index) => index);
    const random = new Sha256ByteStream(this, revealedServerSeed, clientSeed, nonce);
    for (let index = tiles.length - 1; index > 0; index -= 1) {
      const swapIndex = await random.nextInt(index + 1);
      const tile = tiles[index];
      tiles[index] = tiles[swapIndex];
      tiles[swapIndex] = tile;
    }

    return tiles.slice(0, mineCount).sort((left, right) => left - right);
  }
};

class Sha256ByteStream {
  constructor(fairness, serverSeed, clientSeed, nonce) {
    this.fairness = fairness;
    this.serverSeed = serverSeed;
    this.clientSeed = clientSeed;
    this.nonce = nonce;
    this.block = new Uint8Array(0);
    this.offset = 0;
    this.blockIndex = 0;
  }

  async nextInt(bound) {
    const boundBig = BigInt(bound);
    const rejectionLimit = UINT_RANGE - (UINT_RANGE % boundBig);
    let value;
    do {
      value = await this.nextUnsignedInt();
    } while (value >= rejectionLimit);
    return Number(value % boundBig);
  }

  async nextUnsignedInt() {
    if (this.offset + 4 > this.block.length) {
      const message = `${this.clientSeed}:${this.serverSeed}:${this.nonce}:${this.blockIndex}`;
      this.blockIndex += 1;
      this.block = await this.fairness.sha256(message);
      this.offset = 0;
    }
    const value = unsignedInt32(this.block, this.offset);
    this.offset += 4;
    return value;
  }
}

function unsignedInt32(bytes, offset) {
  return BigInt(
    ((bytes[offset] << 24) | (bytes[offset + 1] << 16) | (bytes[offset + 2] << 8) | bytes[offset + 3]) >>> 0
  );
}

export default CasinoFairness;
